Add support for the Satochip hardware wallet

Toporin opened the pull request:
This is a pull request to add support for the Satochip hardware wallet.

The Satochip hardware wallet is based on a jacavard smartcard and is fully open-source.
The wallet is composed of a javacard applet (https://github.com/Toporin/SatochipApplet) that is to be loaded on the smartcard, and an Electron Cash client plugin (https://github.com/Toporin/Electron-Cash-Satochip) that acts as the interface between the card and the network.
The (optional) Satochip-2FA plugin allows to confirm transaction requests on a second android device.

The three-dot diff in the pull request indicates 305 file changed but there is actually much less changes. The two-dot diff is available on this link and list 18 files: https://github.com/Electron-Cash/Electron-Cash/compare/master..Toporin:electron-cash-satochip-pr4.0.11-0.8

More info:
https://github.com/Toporin/ (official repository)
https://prezi.com/p/mpq-xhh3mxjl/satochip-gent-meetup/ (Slides from previous meetups in Belgium)

cculianu commented:
Wow, thanks for this.

I'm currently on vacation and will be back in action Oct. 2 (Tuesday of next week) -- I would like to review this carefully though but it might have to wait until then from my end.

A few thoughts:

Is there any chance you guys can rebase off the latest master? It seems this diff contains stuff we already have in master -- but your commits added them AGAIN somehow. I think your git history got a little out of whack with respect to our history (I suspect you guys copied files over rather than cherry-picking and re-committed them?).

Is there anything you can do to make this pull request smaller somehow? (You can always fix your branch then force push to your `Toporin/electron-cash-satoship-p4xxx` branch to update this pull request).

Toporin commented:
Thanks for your interest in our plugin!
I will try to fix the issue with the pull request and clean it up asap!




On Mon, 23 Sep 2019, 22:28 Calin Culianu, <notifications@github.com> wrote:

> *@cculianu* requested changes on this pull request.
>
> This pull request contains a lot of changes already in master. I suspect
> you started off a much earlier commit as "base" and then you copied files
> back to keep the repo in synch. You should have been doing it via rebase...
> :/
>
> Is there any way you can rebase off latest master -- or otherwise make
> this pull request *only* contain your changes?
>
> I can't merge this as-is..
>
> If you aren't sure how to do that or don't know how -- no worries. I can
> (next week) close this PR and submit another one with just your additions
> in it.
>
> I am really excited you guys did this and can't wait to properly review it
> when I get back from vacation but it would be great if you could make it a
> proper pull request with only your changes contained inside of it...
>
> Thanks!
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Electron-Cash/Electron-Cash/pull/1629?email_source=notifications&email_token=ACSTI2C242ICLF23LB45JADQLERGVA5CNFSM4IZQ3IS2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCFUBL2I#pullrequestreview-292034025>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACSTI2DKHZTIMDW66H44T43QLERGVANCNFSM4IZQ3ISQ>
> .
>

cculianu commented on contrib/build-wine/deterministic.spec:
```diff
@@ -10,7 +10,7 @@ for i, x in enumerate(sys.argv):
     raise BaseException('no name')
 
-
+PYHOME = 'c:/python3.6.8'
```
Hmm. Any chance you can figure out how to do this without hard-coding `3.6.8`?  We just recently got away from the nightmare of having hard-coded versions of everything (including python). It used to be you'd have to edit 4,000 files every time we switch python version and you'd always miss one.  We don't want to slowly decay back to that level of entropy..  Any chance you can glob the dir name from `c:\\python*` or something?

Toporin replied:
I found an easy way to remove the hard-coded version number. I will provide the change in a commit this week.

cculianu commented on contrib/build-wine/deterministic.spec:
```diff
@@ -50,6 +52,7 @@ datas = [
     (home+'lib/locale', 'electroncash/locale'),
     (home+'gui/qt/data/ecsupplemental_win.ttf', 'electroncash_gui/qt/data'),
     (home+'plugins', 'electroncash_plugins'),
+    (PYHOME+'\\Lib\\site-packages\\smartcard\\scard\\_scard.cp36-win32.pyd', '.\\smartcard\\scard\\'), #Satochip
```
Is this needed? PyInstaller doesn't pick it up huh?

Toporin replied:
The maintainer of the project does not provide official releases for Windows 32-bit apparently. Since the project is open-source, it should be possible to build it from sources but it looks quite complex (LudovicRousseau/pyscard#55). The link provided in the script comes from the appveyor account for pyscard (https://ci.appveyor.com/project/LudovicRousseau/pyscard), and is the suggested workaround provided by the maintainer (see LudovicRousseau/pyscard#55 (comment)). Normally, it should be more straightforward for linux/mac.

cculianu commented:
I noticed you guys added Windows install hooks.

What about Linux (AppImage) and macOS builds?  Do you guys only support Windows at the present time with your binary package?

If so we have to basically guard against that and not offer this on Mac & Linux.

If not we have to update the macOS build system and the Linux AppImage builder to pull in the `.pyd` (as you did for Windows already).

Toporin commented:
> I noticed you guys added Windows install hooks.
> 
> What about Linux (AppImage) and macOS builds? Do you guys only support Windows at the present time with your binary package?
> 
> If so we have to basically guard against that and not offer this on Mac & Linux.
> 
> If not we have to update the macOS build system and the Linux AppImage builder to pull in the `.pyd` (as you did for Windows already).

Currently there is only binary for wndows, but I'm working on AppImage and if all goes well, I will commit something end of this week. For macOS I will also have a look, but it might take more time.

Toporin commented:
Just added the integration of Satochip in the appImage. 
Communication with the smartcard requires the pyscard module. To build this package from source requires Swig and Libpcsclite which were added in the docker image. 

Things get a little more complicated for Windows since building pyscard from source requires a rather complex Windows toolchain (swig for Windows, visual c redistributable...). The pyscard maintainer provides pyscard binaries for windows so I added these on the windows build script.

Since pyscard is built a bit differently on windows and linux, I had to make a distinct requirement file ('requirements-satochip.txt'). 

I could not test integration with macOS yet (don't have the necessary hw), but it should be similar to linux since pyscard binaries are provided for macOS on pypi.

cculianu commented:
Hi @Toporin -- I finally got the smart card reader! I'm getting ready to play with this.

I just looked at your build system updates.  They look good!  I really appreciate your taking the time to understand and correctly modify our (rather complex) build system.  Nice job!

I am now going to checkout your branch and begin playing with it to see how it works and/or if there's any suggestions to be offered for your code.

We're getting ready to do a release of Electron Cash (4.0.11) very soon and I'd like very much for this to make it into that release.

I'll see if I can add macOS support if it's not too tough.  I have to first familiarize myself of the hw/sw layer between smartcards and python....

cculianu commented:
Hey @Toporin .. I'm closing this PR, because it already has messy conflicts with the base branch.

I did however take this branch of yours and rebase it manually and resolve conflicts.  I also cleaned up the code slightly to conform to our `PrintError` api for less boiler-platey printing of debug messages.

Would you mind forking off this branch to continue working: https://github.com/Electron-Cash/Electron-Cash/tree/satochip_wip ?

I'm still having trouble talking to the card and/or using the plugin.  It throws errors about not able to talk to card or something.

Am going to try Windows 10 now (I was on macOS -- pyscard works on macos).

Let me know if Linux or Win10 works the best.

cculianu commented:
Error on Windows10 (virtual machine). On Win7 (real box) I couldn't even talk to the smardcard reader.. :(

```
| 45.273| |00| [LogCardConnectionObserver] connecting to VMware Virtual USB CCID 0
| 45.274| |00| [CardConnector] Error during connection: CardConnectionException('Unable to connect with protocol: T0 or T1. The smart card is not responding to a reset. ',) Traceback (most recent call last):
  File "C:\Users\calin\AppData\Local\Temp\_MEI24762\electroncash_plugins\satochip\CardConnector.py", line 95, in __init__
    self.cardservice.connection.connect()
  File "site-packages\smartcard\CardConnectionDecorator.py", line 54, in connect
  File "site-packages\smartcard\pcsc\PCSCCardConnection.py", line 126, in connect
smartcard.Exceptions.CardConnectionException: Unable to connect with protocol: T0 or T1. The smart card is not responding to a reset.
```

On Linux:

```
| 50.561| |00| [satochip] Error during connection: EstablishContextException(2148532253) 
Traceback (most recent call last):
  File "/home/calin/tmp/Electron-Cash/plugins/satochip/satochip.py", line 440, in detect_smartcard_reader
    cardrequest = CardRequest(timeout=5, cardType=self.cardtype)
  File "/home/calin/.local/lib/python3.7/site-packages/smartcard/CardRequest.py", line 61, in __init__
    timeout)
  File "/home/calin/.local/lib/python3.7/site-packages/smartcard/pcsc/PCSCCardRequest.py", line 82, in __init__
    self.hcontext = PCSCContext().getContext()
  File "/home/calin/.local/lib/python3.7/site-packages/smartcard/pcsc/PCSCContext.py", line 56, in __init__
    self.renewContext()
  File "/home/calin/.local/lib/python3.7/site-packages/smartcard/pcsc/PCSCContext.py", line 70, in renewContext
    PCSCContext.instance = PCSCContext.__PCSCContextSingleton()
  File "/home/calin/.local/lib/python3.7/site-packages/smartcard/pcsc/PCSCContext.py", line 40, in __init__
    raise EstablishContextException(hresult)
smartcard.pcsc.PCSCExceptions.EstablishContextException: 'Failure to establish context: Service not available.'
```

I'm actually using VMWare in both cases.  Usually USB devices work over VMWare. 

on macOS (native) I get:

```
| 17.517| |00| [satochip] create_client(): try...
| 17.517| |00| [SatochipClient] __init__()
| 17.518| |00| [RemovalObserver] +Inserted:  
| 17.518| |00| [LogCardConnectionObserver] connecting to Generic Smart Card Reader Interface
| 17.519| |00| [CardConnector] Error during connection: NoCardException('Unable to connect',) Traceback (most recent call last):
  File "/Users/calin/src/ec/electron-cash/plugins/satochip/CardConnector.py", line 95, in __init__
    self.cardservice.connection.connect()
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/smartcard/CardConnectionDecorator.py", line 54, in connect
    self.component.connect(protocol, mode, disposition)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/smartcard/pcsc/PCSCCardConnection.py", line 121, in connect
    raise NoCardException('Unable to connect', hresult=hresult)
smartcard.Exceptions.NoCardException: Unable to connect: No smart card inserted. (0x-7FEFFFF4)

| 17.519| |00| [SatochipClient] __init__(): cc.card_get_ATR()
| 17.519| |00| [satochip] create_client(): exception:Card not connected
| 17.519| |00| [DeviceMgr] scanning devices...
```

cculianu commented:
Note on macOS: if there's no satochip card inserted it blocks polling saying "no card in 5s" (your message).  I insert the card then I get the above error on macOS.

Toporin commented:
When using a virtual machine, you may have to allow the vm to access the card reader. There are two modes for access from vm: 'shared mode' (preferred) and 'USB passthrough' (see https://pubs.vmware.com/fusion-5/index.jsp?topic=%2Fcom.vmware.fusion.help.doc%2FGUID-1B8DDC8A-7372-48A3-9E29-4F93FC6A0F24.html). I have never tested with a macOS host, so I am not entirely sure vmware supports this configuration. 

For debugging purpose and to determine whether the issue comes from python/pyscard or from drivers, you can use the global platform pro tool (https://github.com/martinpaljak/GlobalPlatformPro). Just connect the reader with the smartcard and type `java -jar gp.jar -info` (or `gp.exe -info` on windows) on a command line. If it works, that means the os can access the smartcard.

cculianu commented:
Thanks for the info.

Ok, so I tried macOS native, and Windows under a VM.

I think the reader device itself is being accessed ok on both platforms.  

With no card in I get:

```
$ ./gp.exe -i
GlobalPlatformPro 19.06.16-0-gbaccf34
Running on Windows 10 10.0 amd64, Java 1.8.0_231 by Oracle Corporation
No smart card readers with a card found
```

(same on macOS)

With the Satochip card I get:

```
$ ./gp.exe -i
GlobalPlatformPro 19.06.16-0-gbaccf34
Running on Windows 10 10.0 amd64, Java 1.8.0_231 by Oracle Corporation
Could not connect to Generic Smart Card Reader Interface 0: SCARD_W_UNRESPONSIVE_CARD
```

I get a similar error on macOS: 

```
calin@CalinMBP2019 ~/src/ec/electron-cash $ java -jar gp.jar -i
GlobalPlatformPro 19.06.16-0-gbaccf34
Running on Mac OS X 10.14.6 x86_64, Java 1.8.0_152 by Oracle Corporation
Unlimited crypto policy is NOT installed!
Could not connect to Generic Smart Card Reader Interface: SCARD_E_NO_SMARTCARD
```

As an experiment I tried multiple  bank cards on both Windows VM and macOS native... same error.

Should I have seen info for either the Bank cards or the Satochip card?

It's possible this reader is defective...? Or?

I'm pretty sure macOS native is talking to the reader.  I get debug into about card insertion and removal from your plugin.  I also think Windows is talking to it -- but maybe the reader itself is defective? Hmm..

----

UPDATE: WOAH. I just tried it again on macOS and it worked!  It turns out this reader is a piece of crap and you have to be super gentle with how you insert the card.  This is what I see on macOS:

```
calin@CalinMBP2019 ~/src/ec/electron-cash $ java -jar gp.jar -i
GlobalPlatformPro 19.06.16-0-gbaccf34
Running on Mac OS X 10.14.6 x86_64, Java 1.8.0_152 by Oracle Corporation
Unlimited crypto policy is NOT installed!
Reader: Generic Smart Card Reader Interface
ATR: 3BF81300008131FE454A434F5076323431B7
More information about your card:
    http://smartcard-atr.appspot.com/parse?ATR=3BF81300008131FE454A434F5076323431B7

CPLC: ICFabricator=4790
      ICType=5183
      OperatingSystemID=4791
      OperatingSystemReleaseDate=2082 (2012-03-22)
      OperatingSystemReleaseLevel=3D00
      ICFabricationDate=5188 (2015-07-07)
      ICSerialNumber=01288098
      ICBatchIdentifier=0001
      ICModuleFabricator=4812
      ICModulePackagingDate=5195 (2015-07-14)
      ICCManufacturer=0000
      ICEmbeddingDate=0000 (2010-01-01)
      ICPrePersonalizer=0432
      ICPrePersonalizationEquipmentDate=3031 (2013-01-31)
      ICPrePersonalizationEquipmentID=32383830
      ICPersonalizer=0000
      ICPersonalizationDate=0000 (2010-01-01)
      ICPersonalizationEquipmentID=00000000

Card Data: 
Tag 6: 1.2.840.114283.1
-> Global Platform card
Tag 60: 1.2.840.114283.2.2.2
-> GP Version: 2.2
Tag 63: 1.2.840.114283.3
Tag 64: 1.2.840.114283.4.2.21
-> GP SCP02 i=15
Tag 65: 1.3.656.840.100.2.1.3
Tag 66: 1.3.6.1.4.1.42.2.110.1.2
-> JavaCard v2
Card Capabilities: 
Version: 255 (0xFF) ID:   1 (0x01) type: DES3 length:  16 
Version: 255 (0xFF) ID:   2 (0x02) type: DES3 length:  16 
Version: 255 (0xFF) ID:   3 (0x03) type: DES3 length:  16 
Key version suggests factory keys
```

And on Windows VM:

```
$ ./gp.exe -i
GlobalPlatformPro 19.06.16-0-gbaccf34
Running on Windows 10 10.0 amd64, Java 1.8.0_231 by Oracle Corporation
Reader: VMware Virtual USB CCID 0
ATR: 3BF81300008131FE454A434F5076323431B7
More information about your card:
    http://smartcard-atr.appspot.com/parse?ATR=3BF81300008131FE454A434F5076323431B7

CPLC: ICFabricator=4790
      ICType=5183
      OperatingSystemID=4791
      OperatingSystemReleaseDate=2082 (2012-03-22)
      OperatingSystemReleaseLevel=3D00
      ICFabricationDate=5188 (2015-07-07)
      ICSerialNumber=01288098
      ICBatchIdentifier=0001
      ICModuleFabricator=4812
      ICModulePackagingDate=5195 (2015-07-14)
      ICCManufacturer=0000
      ICEmbeddingDate=0000 (2010-01-01)
      ICPrePersonalizer=0432
      ICPrePersonalizationEquipmentDate=3031 (2013-01-31)
      ICPrePersonalizationEquipmentID=32383830
      ICPersonalizer=0000
      ICPersonalizationDate=0000 (2010-01-01)
      ICPersonalizationEquipmentID=00000000

Card Data:
Tag 6: 1.2.840.114283.1
-> Global Platform card
Tag 60: 1.2.840.114283.2.2.2
-> GP Version: 2.2
Tag 63: 1.2.840.114283.3
Tag 64: 1.2.840.114283.4.2.21
-> GP SCP02 i=15
Tag 65: 1.3.656.840.100.2.1.3
Tag 66: 1.3.6.1.4.1.42.2.110.1.2
-> JavaCard v2
Card Capabilities:
Version: 255 (0xFF) ID:   1 (0x01) type: DES3 length:  16
Version: 255 (0xFF) ID:   2 (0x02) type: DES3 length:  16
Version: 255 (0xFF) ID:   3 (0x03) type: DES3 length:  16
Key version suggests factory keys

```

HA! Thanks a lot for this GP tool -- it helps me troubleshoot! 

I'll go ahead and play with your plugin now...

cculianu commented:
Holy #$#% it works!

I created a Satochip wallet on windows 10!!!

---

<img width="1210" alt="Screen Shot 2019-11-07 at 2 36 47 PM" src="https://user-images.githubusercontent.com/266627/68389632-17771f80-016c-11ea-8fbd-000e38822e58.png">
